### PR TITLE
Cloud 2747 add platform check to doctor

### PR DESCRIPTION
**Pull request by maxres-fr**
alright so this is both operator and platform checks.

A few things that came up when working through this:
1. I think checks might need a failure message. e.g. right now I have a check for the FQDN to not contain example.com. The cli just reports as not healthy. It's just a warning that you're using the example domain.
2. Same as before for the forgerock-sac, there should be a check for the secretsManager != none, because that's also potentially dangerous
3. I wanted to do a check for backups enabled but I don't think that's feasible at the moment.

### CODE REVIEW

**Comment by maxres-fr** [comment]
here's the report for nightly:

image

**Comment by maxres-fr** [comment]
We could potentially make a separate health definition be "misconfigurations"

File: AUTHORS.md [hunk 1]
```diff
@@ -3,7 +3,7 @@
 ## Development Lead
 
 - Max Resnick [maxres-fr](https://github.com/maxres-fr)
-- Lee Baines  [lee-bains](https://github.com/lee-baines)
+- Lee Baines  [lee-baines](https://github.com/lee-baines)
 - Jhonattan Castillo [jrcast](https://github.com/jrcast)
 - Warren Strange [wstrange](https://google.com/wstrang)
```
Comment (wstrange) [inline AUTHORS.md:8]: well, I have zero contributions to the cli, but heck I'll take some credit

**Review by wstrange — APPROVED** [review]
So this uses some rules to drive the checks?

Pretty cool

Nice work

**Comment by maxres-fr** [comment]
**So this uses some rules to drive the checks?**

Yes, it's a boolean expression on any field in the object. The syntax is https://github.com/antonmedv/expr/blob/master/docs/Language-Definition.md

File: AUTHORS.md (hunk 1 above)
Comment (dgoldssfo) [reply AUTHORS.md:8]: Maybe https://github.com/wstrange

File: cmd/doctor.go [hunk 2]
```diff
@@ -6,93 +6,129 @@ import (
...
+		},
+	}
+
+	doctorCmd = &cobra.Command{
+		Use:               "doctor",
+		Aliases:           []string{"dr"},
+		DisableAutoGenTag: true,
+		Short:             "Diagnose common cluster and platform deployments",
+		Long: `
+		Diagnose issues related to running and deploying the ForgeRock platform.
```
Comment (dgoldssfo) [inline cmd/doctor.go:97]: Our TMed product name is "ForgeRock Identity Platform". I found "ForgeRock platform" scattered throughout our READMEs when I was tweaking them the last couple of weeks. Could you please fix this here and elsewhere?

You can abbreviate it as "the platform" once you've used the TMed name.

File: docs/forgeops_doctor_operators.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  -A, --all-namespaces   Default: true. If present, list the requested object(s) across all namespaces. Namespace in current context is ignored even if specified with --namespace. (default true)
+  -h, --help             help for operators
+```
+
+### Options inherited from parent commands
+
+```
+      --as string                      Username to impersonate for the operation
+      --as-group stringArray           Group to impersonate for the operation, this flag can be repeated to specify multiple groups.
+      --cache-dir string               Default cache directory (default "/home/max/.kube/cache")
```
Comment (dgoldssfo) [inline docs/forgeops_doctor_operators.md:38]: s/max/user/ or something like that

**Review by dgoldssfo — APPROVED** [review]
LGTM

File: docs/forgeops_doctor_operators.md (hunk 3 above)
Comment (maxres-fr) [reply docs/forgeops_doctor_operators.md:38]: We have a ticket for this... it's kind of annoying - We just run the "cobra" libs doc method, and that's what it's spitting out.

File: cmd/doctor.go [hunk 4]
```diff
@@ -6,93 +6,129 @@ import (
...
+		Checks that the ForgeRock platform is running.
+	    `,
+		Example: `
+		# validate the platform is running in the current namespace
+		forgeops doctor platform
+		# validate the platform is running in the "prod" namespace
+		forgeops doctor platform -n prod
+		`,
+		DisableAutoGenTag: true,
+		SilenceUsage:      true,
```
Comment (jrcast) [inline cmd/doctor.go:54]:
```suggestion
		SilenceUsage:      true,
		SilenceErrors:     true,
```

File: cmd/doctor.go [hunk 5]
```diff
@@ -6,93 +6,129 @@ import (
...
-		printer.Warnln("Not implemented")
-		return nil
-	},
-	DisableAutoGenTag: true,
-}
+		RunE: func(cmd *cobra.Command, args []string) error {
+			printer.Warnln("Not implemented")
+			return nil
+		},
+		Hidden:            true,
```
Comment (jrcast) [inline cmd/doctor.go:37]: So is this one supposed to be hidden? Just checking.

File: cmd/doctor.go [hunk 6]
```diff
@@ -6,93 +6,129 @@ import (
...
+	operators = &cobra.Command{
+		Use:     "operators",
+		Aliases: []string{"op", "operator"},
+		Short:   "Verify that operators are installed and ready",
+		Long: `
+	    Checks to ensure that required operators are installed and ready.
+	    `,
+		DisableAutoGenTag: true,
+		SilenceErrors:     true,
+		SilenceUsage:      true,
```
Comment (jrcast) [inline cmd/doctor.go:73]: 👍

File: cmd/doctor.go [hunk 7]
```diff
@@ -6,93 +6,129 @@ import (
...
+				return err
+			}
+			return health.Run(clientFactory, hlth, allNamespaces)
+		},
+	}
+
+	doctorCmd = &cobra.Command{
+		Use:               "doctor",
+		Aliases:           []string{"dr"},
+		DisableAutoGenTag: true,
```
Comment (jrcast) [inline cmd/doctor.go:94]:
```suggestion
		DisableAutoGenTag: true,
		SilenceErrors:     true,
		SilenceUsage:      true,
```

File: pkg/health/health.go [hunk 8]
```diff
@@ -52,12 +52,20 @@ func conditionExpression(expression string) k8s.ConditionFunction {
 
 // Check run wait on resource until expression passes
 // note at the moment it doesn't track any success/fails
-func (r *Resource) Check(clientMgr k8s.ClientMgr, namespace string) (bool, error) {
+// a resource is checked in a namespace with the following priority
+// 1. Namespace on resource
+// 2. if 1. is nil then namespace on hlth
+// 3. if 2. is nil
```
Comment (jrcast) [inline pkg/health/health.go:58]: 3. seems to be incomplete. We might not need this one

**Review by jrcast — APPROVED** [review]
LGTM.  I added some comments

File: cmd/doctor.go (hunk 5 above)
Comment (maxres-fr) [reply cmd/doctor.go:37]: At the moment yes.

**Pull request merged**